pkg/blobinfocache/memory/memory.go

```diff
@@ -121,23 +121,37 @@ func (mem *cache) RecordDigestCompressorName(blobDigest digest.Digest, compresso
 }
 
 // appendReplacementCandidates creates prioritize.CandidateWithTime values for (transport, scope, digest), and returns the result of appending them to candidates.
-func (mem *cache) appendReplacementCandidates(candidates []prioritize.CandidateWithTime, transport types.ImageTransport, scope types.BICTransportScope, digest digest.Digest, requireCompressionInfo bool) []prioritize.CandidateWithTime {
-	locations := mem.knownLocations[locationKey{transport: transport.Name(), scope: scope, blobDigest: digest}] // nil if not present
-	for l, t := range locations {
-		compressorName, compressorKnown := mem.compressors[digest]
-		if !compressorKnown {
-			if requireCompressionInfo {
-				continue
-			}
+func (mem *cache) appendReplacementCandidates(candidates []prioritize.CandidateWithTime, transport types.ImageTransport, scope *types.BICTransportScope, digest digest.Digest, requireCompressionInfo bool) []prioritize.CandidateWithTime {
+	compressorName, compressorKnown := mem.compressors[digest]
+	if compressorName == "" {
+		compressorName = blobinfocache.UnknownCompression
+	}
+	if !compressorKnown {
+		if !requireCompressionInfo {
 			compressorName = blobinfocache.UnknownCompression
 		}
+	}
```

Comment:
This is still not equivalent to the BoltDB code.

Also, reading `compressorName` and _only afterwards_ checking `compressorKnown` is inconsistent. If `…Known` is worth checking, it’s worth checking before accessing the value.

Reply:
Pushed a new version where this should look now similar to boltdb implementation PTAL.